plugin search

oulrich1 opened the pull request:
if plugin not found in ./plugs nor packaged in site-packages, then
if the plugin name is one word name try to find under expected
'lobot.plugins' under site-packages

oulrich1 committed:
plugin search

if plugin not found in ./plugs nor packaged in site-packages, then
if the plugin name is one word name try to find under expected
'lobot.plugins' under site-packages

oulrich1 commented:
Just a thought. Although sys.path stuff might be the way to go instead?

lavignes commented on lobot/plugin_manager.py:
```diff
@@ -42,11 +42,24 @@ def find_attributes(self, plugin: Plugin, attribute: str) -> List[_Listener]:
+            try:
+                module = importlib.import_module(module_path)
+            except ImportError:
+                print("> Could not find plugin: '" + str(module_path) + "' in plugs/ nor sys.path.")
```
Id get rid of the print statement. They block the main thread. lobot should only use asyncio. It doesn't have logging yet tho. Issue?

oulrich1 replied:
I agree, it doesnt need the print statements

oulrich1 committed:
Update plugin_manager.py

although maybe the deeper except should just throw the Import Error which would be like normal behavior

lavignes commented on lobot/plugin_manager.py:
```diff
@@ -42,11 +42,24 @@ def find_attributes(self, plugin: Plugin, attribute: str) -> List[_Listener]:
+                    packaged_plugins_dir = "lobot.plugins."
+                    print("> Trying module '" + module_path + "' as '" + packaged_plugins_dir + module_path + "'.")
+                    module_path = packaged_plugins_dir + module_path
+                    module = importlib.import_module(module_path)
```
While we are at it. We should run the module wrapper after loading the module:

``` python
wrapped = Module(module)
```

If it doesn't have any plugins `len(wrapped.plugins) == 0` then it should be the same as failing. Or we could have the Module initializer raise ImportError if the plugins list is empty.

pyrated commented:
No. importlib is the new way. Adding things to the path is messy.

oulrich1 commented:
That makes sense. By the way the module does get wrapped anyways. Right now the 'load_module' only has one return path it doesn't throw out (or shouldn't, unless importlib.reload throws?)

Isn't the wrapped.plugins always going to be of size 1? i think if the pluginmanager has 0 plugins then it should fail as if no plugins were loaded:
`len(plugin_manager.plugins) == 0`

pyrated commented:
Nah. a user could load 0 plugins. That seems fine.

oulrich1 commented:
Ok :+1:

pyrated commented:
The module gets wrapped anyway. But your code could do more and be shorter if you refactor it a bit

oulrich1 commented:
Thats true

oulrich1 committed:
Update plugin_manager.py

Made it simpler

oulrich1 commented:
Thats not right either since it still should be able to load one module locally.. which means it needs to try at least once if its a one word plugin.. *sigh*

oulrich1 committed:
Update plugin_manager.py

pyrated commented:
Fixed by b810f50

pyrated commented:
It is still worth noting that plugin config is still keyed off of the full path... I won't be changing that since it would create ambiguity. I'm actually now not sure If I want relative import support. Namespaces are good.

oulrich1 commented:
That makes sense :)
On Mar 29, 2016 12:21 AM, "Scott LaVigne" <notifications@github.com> wrote:

> It is still worth noting that plugin config is still keyed off of the full
> path... I won't be changing that since it would create ambiguity. I'm
> actually now not sure If I want relative import support. Namespaces are
> good.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/pyrated/lobot/pull/1#issuecomment-202752252>
>
